packages/core/src/StateMachine.ts

```diff
@@ -584,13 +584,95 @@ export class StateMachine<
       children[actorId] = actorRef;
     });
 
+    function resolveStateNodeFromIdOrPath(
+      idOrPath: string,
+      root: StateNode<any, any>
+    ) {
+      try {
+        return root.machine.getStateNodeById(idOrPath);
+      } catch {
+        try {
+          return getStateNodeByPath(root, idOrPath.split('.'));
+        } catch {
+          if (isDevelopment) {
+            console.warn(
+              `Could not resolve StateNode for id/path: ${idOrPath}`
+            );
+          }
+          return idOrPath;
+        }
+      }
+    }
+
+    function reviveHistoryValue(
+      historyValue: unknown,
+      root: StateNode<any, any>,
+      seen = new WeakSet()
```

Comment:
what do we need a recursion for? a `historyValue` value is defined as `Record<string, Array<StateNode<TContext, TEvent>>>` This only requires a single iteration - for each key we should "resolve" all (previously) serialized array items. Those should point to specific state nodes and that's it.

Note that the `historyValue` should get its designated serialization path in `getPersistedSnapshot` within `packages/core/src/State.ts`

Reply:
I have refactored the implementation, removed the recursion, handle invalid state (get rid of it during reconstruction) and featured serialization of historyValue in State, as if I understood correctly, we need only StateNode IDs for correct reconstruction - StateNode itself is not holding any content that needs to be persisted.